bookdown/05-extending/05-extending-mlr3pipelines.Rmd

```diff
@@ -1,5 +1,470 @@
-## Extending mlr3pipelines {#extending-mlr3pipelines}
+1## Extending mlr3pipelines {#extending-mlr3pipelines}
 
 ```{block, type='warning'}
 Work in progress :)
 ```
+
+This tutorial showcases how the `mlr3pipelines` package can be extended to include custom `PipeOps`.
+To run the following examples, we will need a `Task`; we are using the well-known "Iris" task:
+
+```{r 05-extending-015}
+library(mlr3)
+task = mlr_tasks$get("iris")
+task$data()
+```
+
+`mlr3pipelines` is fundamentally built around [`R6`](https://r6.r-lib.org/). When planning to create custom `PipeOp` objects, it can only help to [familiarize yourself with it](https://adv-r.hadley.nz/r6.html).
+
+In principle, all a `PipeOp` must do is inherit from the `PipeOp` R6 class and implement the `train()` and `predict()` functions.
+There are, however, several auxiliary subclasses that can make the creation of *certain* operations much easier.
+
+### General Case Example: `PipeOpCopy` {#ext-pipeopcopy}
+
+A very simple yet useful `PipeOp` is `PipeOpCopy`, which takes a single input and creates a variable number of output channels, all of which receive a copy of the input data.
+It is a simple example that showcases the important steps in defining a custom `PipeOp`.
+We will show a simplified version here, **`PipeOpCopyTwo`**, that creates exactly two copies of its input data.
+
+The following figure visualizes how our `PipeOp` is situated in the `Pipeline` and the significant in- and outputs.
+
+![Pipeop Copy2](images/po_multi_viz.png){ width=90% }
```

Comment:
Does this work in the mlr3book? Looks like some javascript magic (the width setting)

Reply:
Seems to work